magma/backend/mlir/build_magma_graph.py

```diff
@@ -173,11 +174,25 @@ def _visit_inputs(
         )
 
 
+def _check_no_when_loops(graph: Graph):
+    """Temporary guard against https://github.com/phanrahan/magma/issues/1248"""
+    for node in graph.nodes():
+        if not getattr(node, '_is_when_builder_', False):
+            continue
+        for predecessor in graph.predecessors(node):
+            if isinstance(type(node), Register):
+                # broken by register
+                break
+            if predecessor is node:
+                raise NotImplementedError("Dependency loop in when")
+
+
 def build_magma_graph(
         ckt: DefineCircuitKind,
         opts: BuildMagmaGrahOpts = BuildMagmaGrahOpts()) -> Graph:
     ctx = ModuleContext(Graph(), opts)
     _visit_inputs(ctx, ckt, opts.flatten_all_tuples)
     for inst in ckt.instances:
         _visit_inputs(ctx, inst, opts.flatten_all_tuples)
+    _check_no_when_loops(ctx.graph)
```

Comment:
note need a special subclass of `MlirCompilerError` (in `magma/backend/mlir/errors.py`) called `MlirWhenCycleError` (maybe better phrasing is needed :D )

```suggestion
def _check_for_when_cycles(graph: Graph):
    """Temporary guard against https://github.com/phanrahan/magma/issues/1248."""
    for node in graph.nodes():
        if not is_when_builder(node):
            continue
        for predecessor in graph.predecessors(node):
            if isinstance(type(node), Register):  # cycles are broken by registers
                break
            if predecessor is node:
                raise MlirWhenCycleError()


def build_magma_graph(
        ckt: DefineCircuitKind,
        opts: BuildMagmaGrahOpts = BuildMagmaGrahOpts()) -> Graph:
    ctx = ModuleContext(Graph(), opts)
    _visit_inputs(ctx, ckt, opts.flatten_all_tuples)
    for inst in ckt.instances:
        _visit_inputs(ctx, inst, opts.flatten_all_tuples)
    _check_for_when_cycles(ctx.graph)
```